### ISSUE TITLE
Module unarchive - doesn't maintain file permissions from original archive

### ISSUE THREAD

**Issue by edrozenberg** [ansible/ansible#12358]
On Mac:

When using unarchive to unzip a .zip file, the destination output gets "drwxrwxrwx" permissions, which are not the same as the original zip file. If I unzip the original with "unzip" manually, I get the correct permissions.

I'm aware of the "mode=" however doing "mode=u=rwX,g=rX,o=rX" leaves all files with executable bits turned on, whereas "mode=u=rw,g=r,o=r" leaves directories un-enterable, so I'm not able to achieve the desired permissions without a second task to call a command to do another chmod.

**Comment by Kr08ti** [comment]
Hi,
When you unzip a file, the default permission it gets are "drwxr-xr-x", so if you want to give permission which let you enter into the file, you can try mode=766/u=rwx,g=rw,o=rw. The owner of the  file will only have executable permissions and group and and other have read and write.
You can also try mode=522/u=rx,g=w,o=w, or mode=744/u=rwx,g=r,o=r if you want your unziped file to be more protected. 
There is no harm in giving executable permissions to the owner of the file.

Thanks

**Comment by abadger** [comment]
I'm not able to duplicate this behaviour on Linux:

```
$ ansible localhost -m unarchive -a 'src=test.zip dest=/home/badger/.openttd/save/autosave/un'
localhost | success >> {
    "changed": true,
    "check_results": {
        "unarchived": false
    },
    "dest": "/home/badger/.openttd/save/autosave/un",
    "extract_results": {
        "cmd": "/usr/bin/unzip -o \"/home/badger/.ansible/tmp/ansible-tmp-1444445277.26-85464795209723/source\" -d \"/home/badger/.openttd/save/autosave/un\"",
        "err": "",
        "out": "Archive:  /home/badger/.ansible/tmp/ansible-tmp-1444445277.26-85464795209723/source\n   creating: /home/badger/.openttd/save/autosave/un/test/\n extracting: /home/badger/.openttd/save/autosave/un/test/read  \n extracting: /home/badger/.openttd/save/autosave/un/test/execute  \n   creating: /home/badger/.openttd/save/autosave/un/test/directory/\n",
        "rc": 0
    },
    "gid": 1000,
    "group": "badger",
    "handler": "ZipArchive",
    "mode": "0775",
    "owner": "badger",
    "secontext": "unconfined_u:object_r:user_home_t:s0",
    "size": 4096,
    "src": "/home/badger/.ansible/tmp/ansible-tmp-1444445277.26-85464795209723/source",
    "state": "directory",
    "uid": 1000
}
$ ls -al un/test                                                     (19:48:03)
total 12
drwxrwxr-x. 3 badger badger 4096 Oct  9 19:44 .
drwxrwxr-x. 3 badger badger 4096 Oct  9 19:47 ..
drwxrwxr-x. 2 badger badger 4096 Oct  9 19:43 directory
-rwxr-xr-x. 1 badger badger    0 Oct  9 19:43 execute
-r--r--r--. 1 badger badger    0 Oct  9 19:43 read
```

All ansible is doing is:

```
        cmd = '%s -o "%s" -d "%s"' % (self.cmd_path, self.src, self.dest)
```

where cmd_path is the path to unzip that ansible discovered, src is the specified source directory and dest is the specified destination directory.

Perhaps you have more than one "unzip" command on your computer and ansiblee is finding a different one than when you run unzip manually?

Or is this a problem general to Mac OS?  We'll need more information in order to debug this, I'm afraid.

**Comment by abadger** [comment]
part of the unarchive module lives in the main ansible/ansible repo and another part lives in https://github.com/ansible/ansible-modules-core/ . It's highly likely that if ansible can fix this somehow it would be in the part that is in the ansible-modules-core repository but I can't quite be sure yet.  Will need more information before we can figure that out.

**Comment by dagwieers** [comment]
I cannot reproduce this either. Beware that unzip uses the umask value when creating new files/directories. And this may be different when running through Ansible and testing elsewhere. Usually user umask is set to 002 whereas the umask for root is set to 022.

@gregdek I'd say close this issue because we lack the needed feedback.

**Comment by abadger** [comment]
Due to lack of info, closing this.  If you still suffer from this issue, please open a new ticket, hopefully with more information to reproduce the issue.  Thanks.

**Comment by donkeyx** [comment]
Hey guys, 

I am having this exact issue and can give you the tar etc if you need. In this case, i have the tarred directory contents wtih the correct perms : 775 but when copied with unarchive to remote server it will become 755, which breaks my build. As you can see below the unarchived version has reverted to 755 rather than what was in the tar archive.

My ansible code is : 
```ansible
  - name: Copy and extract packaged app to hosts
    unarchive:
      src=../{{ release_subfolder }}.tar.gz
      dest={{ item.documentrootdir }}/releases/{{ release_subfolder }}
      group=apache
    with_items: "{{ apache_vhosts_enabled }}"
```

Source tar contains : 

```bash
drwxrwxr-x go/apache         0 2016-12-16 13:34 ./tmp/
drwxrwxr-x go/apache         0 2016-12-16 13:34 ./tmp/tests/
drwxrwxr-x go/apache         0 2016-12-16 13:34 ./tmp/cache/
drwxrwxr-x go/apache         0 2016-12-16 13:34 ./tmp/cache/persistent/
drwxrwxr-x go/apache         0 2016-12-16 13:34 ./tmp/cache/models/
drwxrwxr-x go/apache         0 2016-12-16 13:34 ./tmp/cache/views/
drwxrwxr-x go/apache         0 2016-12-16 13:34 ./tmp/sessions/
drwxrwxr-x go/go             0 2016-07-20 22:31 ./vendor/mpdf/mpdf/tmp/
-rw-rw-r-- go/go            14 2016-07-20 22:31 ./vendor/mpdf/mpdf/tmp/.gitignore
```

Result on server after unarchive: 

```bash
drwxr-xr-x  5 go apache 4096 Dec 16 13:34 .
drwxrwxr-x 12 go apache 4096 Dec 16 13:34 ..
drwxr-xr-x  5 go apache 4096 Dec 16 13:34 cache
drwxr-xr-x  2 go apache 4096 Dec 16 13:34 sessions
drwxr-xr-x  2 go apache 4096 Dec 16 13:34 tests
```
